### PR TITLE
alpha matching: ensure that extracted header values are not reused between streams

### PR DESCRIPTION
**Pull request by snowp**
Fixes an issue where the extracted header value would be persisted between streams, resulting in incorrect
matching results. Streams that should not match would match due to the previous stream having a header
set that matches the result. 

Risk Level: Medium, change to alpha feature
Testing: UT test case
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by snowp** [comment]
I think perhaps later on it would make sense to have the API return a implementable interface so that we can use that to wrap various string storage objects to avoid unnecessary copies, but this should fix the immediate issue.

File: test/common/http/matching/inputs_test.cc [hunk 1]
```diff
@@ -9,16 +9,22 @@ namespace Envoy {
...
-  data.onRequestHeaders(request_headers);
 
-  EXPECT_EQ(input.get(data).data_, "bar");
-  EXPECT_EQ(input.get(data).data_, "bar");
+  {
+    TestRequestHeaderMapImpl request_headers({{"header", "bar"}});
+    data.onRequestHeaders(request_headers);
+
+    EXPECT_EQ(input.get(data).data_, "bar");
+    EXPECT_EQ(input.get(data).data_, "bar");
```
Comment (dmitri-d) [inline test/common/http/matching/inputs_test.cc:21]: Here and below  -- why the duplicate check is required? Perhaps add a comment here?

**Comment by dmitri-d** [comment]
Could you also add a test for `Matcher::DataInputGetResult get` when there's no requested header in the data? (to verify we get `absl::nullopt` here: https://github.com/envoyproxy/envoy/pull/16721/files#diff-2df46d0706764a215b8c3bf412d40761a2f7c02f8abb798aa929cad394e721b7R44)

**Comment by mattklein123** [comment]
LMK when CI issues are resolved.

/wait

File: test/common/http/matching/inputs_test.cc (hunk 1 above)
Comment (snowp) [reply test/common/http/matching/inputs_test.cc:21]: Old code I think, let me remove one of them

**Review by mattklein123 — COMMENTED** [review]
Thanks looks good. Can you merge main?

/wait

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**